fairlearn/metrics/_metric_frame.py

```diff
@@ -195,24 +209,166 @@ def __init__(self,
+            For cases where no names are provided we generate names ``sensitive_feature_[n]``.
+
+        control_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray, DataFrame
+            Control features are similar to sensitive features, in that they
+            divide the input data into subgroups.
+            Unlike the sensitive features, aggregations are not performed
+            across the control features - for example, the ``overall`` property
+            will have one value for each subgroup in the control feature(s),
+            rather than a single value for the entire data set.
+            Control features can be specified similarly to the sensitive features.
+            However, their default names (if none can be identified in the
+            input values) are of the format ``control_feature_[n]``.
+
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+
+        sample_params : dict
+            Parameters for the metric function(s). If there is only one metric function,
+            then this is a dictionary of strings and array-like objects, which are split
+            alongside the ``y_true`` and ``y_pred`` arrays, and passed to the metric function.
+            If there are multiple metric functions (passed as a dictionary), then this is
+            a nested dictionary, with the first set of string keys identifying the
+            metric function name, with the values being the string-to-array-like dictionaries.
+        """
+        if not self._streaming:
+            raise NotImplementedError("This MetricFrame does not support adding data.")
+        check_consistent_length(y_true, y_pred, sensitive_features)
+
         y_t = _convert_to_ndarray_and_squeeze(y_true)
         y_p = _convert_to_ndarray_and_squeeze(y_pred)
 
-        func_dict = self._process_functions(metrics, sample_params)
+        self._y_true.append(y_t)
+        self._y_pred.append(y_p)
+        self._s_f.append(sensitive_features)
+        if self._c_f is not None:
+            check_consistent_length(y_true, control_features)
+            self._c_f.append(control_features)
+        elif control_features is not None:
+            raise ValueError(_CF_BAD_STATE_ERR)
 
+        if self._s_p is not None:
+            if sample_params is not None:
+                # All arrays should be of equal length
+                check_consistent_length(y_true, *sample_params.values())
+                self._s_p = self._concat_batches([self._s_p, sample_params])
+            else:
+                raise ValueError("MetricFrame expected `sample_params` to be supplied.")
+
+        # Reset metric
+        self._overall = None
+        self._by_group = None
+
+    def _concat_batches(self, batches):
+        """Concatenate a list of items together."""
+        if len(batches) == 0:
+            raise ValueError(_EMPTY_BATCHES_ERR)
```

Comment:
instead of raise, I'd just return None.

Reply:
I'm afraid that this could lead to issues in the future.
Other libraries also raise:

Numpy: `ValueError: need at least one array to concatenate`
Pandas: `ValueError: No objects to concatenate`

Reply:
I'm happy with raising here.